### PR TITLE
MS coco dataset training

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: python/benchmarks/coco/datagen.py [hunk 1]
```diff
@@ -36,6 +44,9 @@ def _instances_to_df(self, split, instances_json):
         category_df = pd.DataFrame(instances_json["categories"]).rename(
             {"id": "category_id"}, axis=1
         )
+        # df["bbox"] = lance.types.Box2dArray(df["bbox"])
```
Comment (changhiskhan) [inline python/benchmarks/coco/datagen.py:47]: remove these lines?

File: python/benchmarks/coco/datagen.py [hunk 2]
```diff
@@ -242,9 +259,25 @@ def _aggregate_annotations(annotations):
...
+    version,
+    fmt,
+    embedded,
+    output_path,
+    group_size: int,
+    max_rows_per_file: int,
+):
     converter = CocoConverter(base_uri, version=version)
     df = converter.read_metadata()
+    # print(df)
```
Comment (changhiskhan) [inline python/benchmarks/coco/datagen.py:273]: remove these lines altogether?

File: python/lance/pytorch/data.py [hunk 3]
```diff
@@ -68,7 +89,7 @@ class LanceDataset(IterableDataset):
     def __init__(
         self,
         root: Union[str, Path],
-        columns: Optional[List[str]] = None,
+        columns: Optional[Union[List[str], Dict[str, str]]] = None,
```
Comment (changhiskhan) [inline python/lance/pytorch/data.py:92]: what's the dict use case?

File: python/lance/pytorch/data.py [hunk 4]
```diff
@@ -55,6 +75,7 @@ def to_tensor(arr: pa.Array) -> Union[torch.Tensor, PIL.Image.Image]:
     elif pa.types.is_string(arr.type) or pa.types.is_large_string(arr.type):
         return np_arr.astype(np.str_)
     else:
+        print(np_arr)
```
Comment (changhiskhan) [inline python/lance/pytorch/data.py:78]: remove print?

File: python/benchmarks/coco/common.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    """General object detection model for training and inference benchmarks.
+
+    Hyper parameters are referred to
+
+    https://github.com/pytorch/vision/blob/main/references/detection
+    """
+
+    def __init__(self, benchmark: str = "train"):
+        super().__init__()
+        self.backbond = torchvision.models.detection.ssd300_vgg16()
```
Comment (changhiskhan) [inline python/benchmarks/coco/common.py:113]: backbone

File: python/benchmarks/coco/common.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+    def training_step(self, batch, batch_idx):
+        if self._benchmark == "io":
+            return
+        images, targets = batch
+        loss_dict = self.backbond(images, targets)
+        loss = sum(loss for loss in loss_dict.values())
+        if torch.isinf(loss) or torch.isnan(loss) and self._benchmark == "train":
+            print(f"Loss is {loss}, exit")
+            sys.exit(1)
```
Comment (changhiskhan) [inline python/benchmarks/coco/common.py:134]: why sys.exit instead of just raising an exception?

File: python/benchmarks/coco/transforms.py [hunk 7]
```diff
@@ -0,0 +1,662 @@
+#!/usr/bin/env python
+#
+# Data augmentation during object detection training.
+#
+# Copied from https://raw.githubusercontent.com/pytorch/vision/main/references/detection/transforms.py
```
Comment (changhiskhan) [inline python/benchmarks/coco/transforms.py:5]: really interesting these are not available as library code

File: python/benchmarks/coco/train.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        dp = IterableWrapper(dataset).shuffle(buffer_size=64)
+        train_loader = torch.utils.data.DataLoader(
+            dp,
+            num_workers=num_workers,
+            batch_size=batch_size,
+            collate_fn=collate_fn,
+            pin_memory=True,
+        )
+    elif data_format == "raw":
+        dataset = RawCocoDataset(uri, transform=get_transform(benchmark))
```
Comment (changhiskhan) [inline python/benchmarks/coco/train.py:141]: do we not need the IterableWrapper for shuffle here?

File: python/benchmarks/coco/train.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+def get_transform(benchmark: str) -> Callable:
+    if benchmark == "io":
+        return prepare_target
+    elif benchmark == "no_augmentation":
+        return T.Compose(
+            [prepare_target, T.PILToTensor(), T.ConvertImageDtype(torch.float)]
+        )
+    elif benchmark == "train":
+        # https://github.com/pytorch/vision/blob/24890d718f5a73586ef093371912b5b37a5b0d46/references/detection/presets.py#L37
+        return T.Compose(
```
Comment (changhiskhan) [inline python/benchmarks/coco/train.py:43]: duplicate?

**Review by changhiskhan — COMMENTED** [review]
curious if the trained model was good?

File: python/benchmarks/coco/transforms.py (hunk 7 above)
Comment (eddyxu) [reply python/benchmarks/coco/transforms.py:5]: Yea, the official torchvision one is transform applied to image only.  But these data augmentation transforms apply to both image and target. for example, flip needs to flip both image and bbox.

File: python/benchmarks/coco/train.py (hunk 8 above)
Comment (eddyxu) [reply python/benchmarks/coco/train.py:141]: This is MapDataset, that DataLoader can shuffle it.

File: python/benchmarks/coco/README.md [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+```sh
+
+./datagen.py -f lance /path/to/coco
+```
+
+## Training
+
+```sh
+
+./train.py -f lance --epoch 10 [--benchmark io] /path/to/coco
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/benchmarks/coco/README.md:14]: Ok

File: python/benchmarks/coco/datagen.py (hunk 2 above)
Comment (eddyxu) [reply python/benchmarks/coco/datagen.py:273]: Done

File: python/benchmarks/coco/datagen.py (hunk 1 above)
Comment (eddyxu) [reply python/benchmarks/coco/datagen.py:47]: Done

File: python/lance/pytorch/data.py (hunk 3 above)
Comment (eddyxu) [reply python/lance/pytorch/data.py:92]: Oh, i was intent to allow `alias` via columns parameter, like `{"labels": "annotations.category_id",  "boxes": "annotations.bbox"}`. 

Will file another ticket to do it appropriately.

File: python/benchmarks/coco/train.py (hunk 9 above)
Comment (eddyxu) [reply python/benchmarks/coco/train.py:43]: Fixed.

File: python/benchmarks/coco/common.py (hunk 6 above)
Comment (eddyxu) [reply python/benchmarks/coco/common.py:134]: Good point. lemme change to an exception.

File: python/benchmarks/coco/common.py (hunk 5 above)
Comment (eddyxu) [reply python/benchmarks/coco/common.py:113]: Fixed

File: python/lance/pytorch/data.py (hunk 4 above)
Comment (eddyxu) [reply python/lance/pytorch/data.py:78]: done

File: python/lance/pytorch/data.py (hunk 3 above)
Comment (eddyxu) [reply python/lance/pytorch/data.py:92]: Filed https://github.com/eto-ai/lance/issues/178

**Review by changhiskhan — APPROVED** [review]
LGTM
